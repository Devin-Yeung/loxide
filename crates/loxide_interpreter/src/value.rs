#[derive(Debug, Clone, PartialEq)]
pub enum Value {
    Number(f64),
    String(String),
    Boolean(bool),
    Nil,
}

impl Value {
    pub fn is_number(&self) -> bool {
        matches!(self, Value::Number(_))
    }

    pub fn is_string(&self) -> bool {
        matches!(self, Value::String(_))
    }

    pub fn is_boolean(&self) -> bool {
        matches!(self, Value::Boolean(_))
    }

    pub fn is_nil(&self) -> bool {
        matches!(self, Value::Nil)
    }

    pub fn as_number(&self) -> Option<f64> {
        match self {
            Value::Number(v) => Some(*v),
            _ => None,
        }
    }

    pub fn as_boolean(&self) -> Option<bool> {
        match self {
            Value::Boolean(v) => Some(*v),
            _ => None,
        }
    }
}
