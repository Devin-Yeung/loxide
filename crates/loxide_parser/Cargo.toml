[package]
name = "loxide_parser"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = { workspace = true, features = ["glob"] }
thiserror = { workspace = true }
miette = { workspace = true }
loxide_diagnostic = { workspace = true }
clap = { version = "4.5.13", features = ["derive"] }

[dev-dependencies]
loxide_macros = { workspace = true }
loxide_testsuite = { workspace = true }
automod = "1.0.14"