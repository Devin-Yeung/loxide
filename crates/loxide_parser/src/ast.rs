#[derive(Debug)]
pub struct Expr<'src> {
    pub kind: ExprKind<'src>,
}

#[derive(Debug)]
pub enum ExprKind<'src> {
    Literal(Literal<'src>),
    Unary(UnaryExpr<'src>),
    Binary(BinaryExpr<'src>),
    Grouped(Box<Expr<'src>>),
    Variable(Variable<'src>),
    Assign(AssignExpr<'src>),
}

#[derive(Debug)]
pub enum Literal<'src> {
    String(std::borrow::Cow<'src, str>),
    Number(f64),
    Boolean(bool),
    Nil,
}

#[derive(Debug)]
pub enum UnaryOperator {
    Minus,
    Bang,
}

#[derive(Debug)]
pub struct UnaryExpr<'src> {
    pub operator: UnaryOperator,
    pub expr: Box<Expr<'src>>,
}

#[derive(Debug)]
pub enum BinaryOperator {
    EqualEqual,
    BangEq,
    GreaterEqual,
    Greater,
    Less,
    LessEqual,
    Minus,
    Plus,
    Slash,
    Star,
}

#[derive(Debug)]
pub struct BinaryExpr<'src> {
    pub lhs: Box<Expr<'src>>,
    pub rhs: Box<Expr<'src>>,
    pub operator: BinaryOperator,
}

#[derive(Debug)]
pub struct GroupedExpr<'src> {
    pub expr: Box<Expr<'src>>,
}

#[derive(Debug)]
pub struct AssignExpr<'src> {
    pub name: Variable<'src>,
    pub value: Box<Expr<'src>>,
}

impl<'src> AssignExpr<'src> {
    pub fn new(name: &'src str, value: Expr<'src>) -> Self {
        AssignExpr {
            name: Variable { name },
            value: Box::new(value),
        }
    }
}

#[derive(Debug)]
pub struct Variable<'src> {
    pub name: &'src str,
}

#[derive(Debug)]
pub struct ConditionStmt<'src> {
    pub condition: Expr<'src>,
    pub then_branch: Box<Stmt<'src>>,
    pub else_branch: Option<Box<Stmt<'src>>>,
}

#[derive(Debug)]
pub struct WhileStmt<'src> {
    pub condition: Expr<'src>,
    pub body: Box<Stmt<'src>>,
}

#[derive(Debug)]
#[non_exhaustive]
pub enum Stmt<'src> {
    Expression(Expr<'src>),
    PrintStmt(Expr<'src>),
    VarDeclaration(Variable<'src>, Option<Expr<'src>>),
    Block(Vec<Stmt<'src>>),
    Condition(ConditionStmt<'src>),
    While(WhileStmt<'src>),
}
