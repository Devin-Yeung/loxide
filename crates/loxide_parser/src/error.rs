use thiserror::Error;

#[derive(Error, Debug, Eq, PartialEq, Copy, Clone)]
pub enum SyntaxError {
    #[error("Unexpected character at line {0}")]
    UnexpectedChar(usize),
    #[error("Unexpected EOF")]
    UnexpectedEOF,
    #[error("Expected")]
    Expect(&'static str),
    #[error("Invalid Conversion")]
    InvalidConversion,
    #[error("Invalid Number")]
    InvalidNumber,
    #[error("Invalid string literal")]
    InvalidStringLiteral,
    #[error("Invalid keyword")]
    InvalidKeywords,
}
