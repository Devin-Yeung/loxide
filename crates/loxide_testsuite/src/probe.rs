use once_cell::sync::Lazy;
use std::backtrace::Backtrace;
use std::collections::HashMap;
use std::fmt::{Debug, Formatter};
use std::sync::Mutex;

static PROBE: Lazy<Mutex<Probe>> = Lazy::new(|| Mutex::new(Probe::new()));
#[derive(Debug)]
pub struct Probe {
    mapping: HashMap<String, Vec<FootPrint>>,
}

#[derive(Clone, PartialEq)]
pub struct FootPrint(String);

impl PartialEq<FootPrint> for &str {
    fn eq(&self, other: &FootPrint) -> bool {
        self.eq(&other.0)
    }
}

impl PartialEq<&str> for FootPrint {
    fn eq(&self, other: &&str) -> bool {
        self.0.eq(other)
    }
}

impl Debug for FootPrint {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.0)
    }
}

impl Probe {
    fn new() -> Self {
        Self {
            mapping: HashMap::new(),
        }
    }

    pub fn register(test_name: &str) {
        PROBE
            .lock()
            .unwrap()
            .mapping
            .insert(test_name.to_string(), Vec::new());
    }

    pub fn probe(content: String) {
        let raw = Backtrace::capture().to_string();
        let backtrace = Self::backtrace(&raw);
        let mut guard = PROBE.lock().unwrap();
        for (k, v) in guard.mapping.iter_mut() {
            if backtrace.iter().any(|func| func == k) {
                v.push(FootPrint(content));
                break;
            }
        }
    }

    pub fn footprints(test_name: &str) -> Vec<FootPrint> {
        let guard = PROBE.lock().unwrap();
        guard.mapping.get(test_name).unwrap_or(&Vec::new()).clone()
    }

    fn backtrace(backtrace: &str) -> Vec<&str> {
        let mut stacks = Vec::new();
        for line in backtrace.lines() {
            let trimmed = line.trim_start();
            if &trimmed[..2] != "at" {
                let func = trimmed[trimmed.find(':').unwrap() + 1..].trim();
                stacks.push(func);
            }
        }
        stacks
    }
}

#[cfg(test)]
mod tests {
    use crate::{footprints, probe, register};

    fn nested(num: i32) {
        outer(num);
    }

    fn outer(num: i32) {
        inner(num);
    }

    fn inner(num: i32) {
        probe!(num);
    }

    #[test]
    fn same_prefix() {
        fn same_prefix_foo() {
            register!();
            nested(42);
            assert_eq!("42", footprints!()[0]);
        }

        fn same_prefix_bar() {
            register!();
            nested(-42);
            assert_eq!(footprints!()[0], "-42");
        }
        same_prefix_foo();
        same_prefix_bar();
    }
}
