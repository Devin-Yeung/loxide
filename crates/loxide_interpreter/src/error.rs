use crate::value::{Value, ValueKind};
use loxide_parser::token::Span;
use miette::Diagnostic;
use thiserror::Error;

#[derive(Debug, Error, Diagnostic)]
#[non_exhaustive]
pub enum RuntimeError {
    #[error("Invalid unary operand, expected: {1}")]
    InvalidUnaryOperand(#[label("expect a {1} type here")] Span, &'static str),
    #[error("Invalid binary operand, expected: {1}")]
    InvalidBinaryOperand(#[label("expect a {1} type here")] Span, &'static str),
    #[error("Incompatible operand, {lhs_ty} is incompatible with {rhs_ty}")]
    IncompatibleBinaryOperand {
        #[label("has type `{lhs_ty}` here")]
        lhs_span: Span,
        lhs_ty: ValueKind,
        #[label("has type `{rhs_ty}` here")]
        rhs_span: Span,
        rhs_ty: ValueKind,
    },
    #[error("Undefined Variable: {1}")]
    UndefinedVariable(#[label("variable `{1}` is not defined")] Span, String),
    #[error("Expected boolean in condition expression")]
    ExpectedBoolean(
        #[label("expect a boolean type here, but got type `{1}`")] Span,
        ValueKind,
    ),
    #[error("Return value")]
    ReturnValue(Value),
    #[error("Bad arity, expected: {expected}, found: {found}")]
    BadArity {
        #[label("function declare here with {expected} parameters")]
        callee_span: Span,
        #[label("function call here with {found} parameters")]
        caller_span: Span,
        expected: usize,
        found: usize,
    },
    #[error("Function call on a non-callable value")]
    CallOnNonCallable(
        #[label("expect a callable value here, but got type `{1}`")] Span,
        ValueKind,
    ),
    #[error("Return in top level")]
    ReturnInTopLevel(#[label("return statement is not allowed in top level")] Span),
    #[error("Too many parameters, allow at most 255, found {1}")]
    TooManyParameters(#[label("Can't have more than 255 parameters")] Span, usize),
    #[error("Too many arguments, allow at most 255, found {1}")]
    TooManyArguments(#[label("Can't have more than 255 arguments")] Span, usize),
}

#[derive(Debug)]
pub(crate) enum PrivateRuntimeError {
    /// Bad Arity Error without Span
    BadArity { expected: usize, found: usize },
    /// Transparent Error of `RuntimeError`
    Transparent(RuntimeError),
}
