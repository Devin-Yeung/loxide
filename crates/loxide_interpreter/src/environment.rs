use crate::error::RuntimeError;
use crate::value::Value;
use std::collections::HashMap;

pub struct Environment<'a> {
    values: HashMap<String, Value>,
    parent: Option<&'a Environment<'a>>,
}

impl<'a> Environment<'a> {
    pub fn new(parent: Option<&'a Environment<'a>>) -> Environment<'a> {
        Environment {
            values: HashMap::new(),
            parent,
        }
    }

    pub fn get(&self, name: &str) -> Result<&Value, RuntimeError> {
        self.values.get(name).map_or_else(
            || {
                self.parent
                    .map(|parent| parent.get(name))
                    .unwrap_or(Err(RuntimeError::UndefinedVariable(name.to_string())))
            },
            Ok,
        )
    }

    pub fn define(&mut self, name: String, value: Value) {
        self.values.insert(name, value);
    }

    pub fn update(&mut self, name: &str, value: Value) -> Result<(), RuntimeError> {
        match self.values.get_mut(name) {
            None => return Err(RuntimeError::UndefinedVariable(name.to_string())),
            Some(existed) => {
                *existed = value;
                Ok(())
            }
        }
    }
}

#[cfg(test)]
mod test {
    use crate::environment::Environment;
    use crate::value::Value;

    #[test]
    fn nested() {
        let mut global = Environment::new(None);
        global.define("a".to_string(), Value::Number(1.0));
        let mut outer = Environment::new(Some(&global));
        outer.define("b".to_string(), Value::Number(2.0));
        outer.define("c".to_string(), Value::Number(0.0));
        let mut inner = Environment::new(Some(&outer));
        inner.define("c".to_string(), Value::Number(3.0)); // shadow c

        assert_eq!(inner.get("a").unwrap().as_number().unwrap(), 1.0);
        assert_eq!(inner.get("b").unwrap().as_number().unwrap(), 2.0);
        assert_eq!(inner.get("c").unwrap().as_number().unwrap(), 3.0);
        assert_eq!(outer.get("c").unwrap().as_number().unwrap(), 0.0);
    }
}
