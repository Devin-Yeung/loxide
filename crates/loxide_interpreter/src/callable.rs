use crate::environment::Environment;
use crate::error::PrivateRuntimeError::Transparent;
use crate::error::{PrivateRuntimeError, RuntimeError};
use crate::eval::Evaluable;
use crate::value::Value;
use loxide_parser::ast::FunDeclaration;
use loxide_parser::token::Span;
use std::io::Write;
use std::sync::Arc;

#[derive(Debug, Clone)]
#[non_exhaustive]
pub enum Callable {
    Function(LoxFunction),
}

impl PartialEq for Callable {
    fn eq(&self, _other: &Self) -> bool {
        todo!()
    }
}

impl Callable {
    pub(crate) fn call(
        &self,
        arguments: Vec<Value>,
        env: &mut Environment,
        stdout: &mut impl Write,
    ) -> Result<Value, PrivateRuntimeError> {
        match self {
            Callable::Function(func) => func.call(arguments, env, stdout),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct LoxFunction {
    // we don't want to introduce any lifetime here,
    // since it will propagate to `Value`
    declaration: Arc<FunDeclaration>,
}

impl LoxFunction {
    pub fn new(declaration: Arc<FunDeclaration>) -> LoxFunction {
        LoxFunction { declaration }
    }

    pub(crate) fn call(
        &self,
        arguments: Vec<Value>,
        env: &mut Environment,
        stdout: &mut impl Write,
    ) -> Result<Value, PrivateRuntimeError> {
        let mut env = env.extend();

        if self.declaration.params.len() != arguments.len() {
            return Err(PrivateRuntimeError::BadArity {
                expected: self.declaration.params.len(),
                found: arguments.len(),
            });
        }

        // binding args to the environment
        for (value, arg) in arguments.into_iter().zip(&self.declaration.params) {
            env.define(arg, value);
        }

        for stmt in &self.declaration.body {
            match stmt.eval(&mut env, stdout) {
                Ok(_) => continue,
                Err(RuntimeError::ReturnValue(ret)) => return Ok(ret),
                Err(e) => return Err(Transparent(e)),
            }
        }

        Ok(Value::Void)
    }
}

impl Callable {
    pub fn function(declaration: Arc<FunDeclaration>) -> Callable {
        Callable::Function(LoxFunction::new(declaration))
    }

    pub fn params_span(&self) -> Span {
        match self {
            Callable::Function(func) => func.declaration.paren_token,
        }
    }
}
