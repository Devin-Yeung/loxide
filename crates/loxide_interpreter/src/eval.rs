use crate::callable::Callable;
use crate::environment::Environment;
use crate::error::{PrivateRuntimeError, RuntimeError};
use crate::value::{Value, ValueKind};
use loxide_parser::ast::{
    AssignExpr, BinaryExpr, BinaryOperator, CallExpr, ConditionStmt, Expr, ExprKind, ForStmt,
    GroupedExpr, Identifier, Literal, ReturnStmt, Stmt, StmtKind, UnaryExpr, UnaryOperator,
    WhileStmt,
};
use loxide_testsuite::probe;
use std::io::Write;

pub trait Evaluable {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError>;
}

macro_rules! inner_or {
    ($val:expr, $ty:tt, $err:expr) => {{
        ::paste::paste! {
            if $val.[<is_$ty>]() {
                Ok($val.[<as_$ty>]().unwrap())
            } else {
                Err($err)
            }
        }
    }};
}

impl Evaluable for Stmt {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        self.kind.eval(env, stdout)
    }
}

impl Evaluable for StmtKind {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        match self {
            StmtKind::Expression(e) => e.eval(env, stdout),
            StmtKind::PrintStmt(e) => {
                let v = e.eval(env, stdout)?;
                let _ = probe!(&v, span => e.span().into());
                writeln!(stdout, "{v}").expect("Failed to write to stdout");
                Ok(Value::Void)
            }
            StmtKind::VarDeclaration(var, expr) => {
                let val = expr.eval(env, stdout)?;
                env.define(var, val);
                Ok(Value::Void)
            }
            StmtKind::FunDeclaration(decl) => {
                let callable = Callable::function(decl.clone(), env.extend());
                env.define(&decl.name, Value::Callable(callable));
                Ok(Value::Void)
            }
            StmtKind::Block(stmts) => {
                // TODO: find a way to test the correctness of this evaluation
                let mut scope = env.extend();
                for stmt in stmts {
                    stmt.eval(&mut scope, stdout)?;
                }
                Ok(Value::Void)
            }
            StmtKind::Condition(cond) => cond.eval(env, stdout),
            StmtKind::While(cond) => cond.eval(env, stdout),
            StmtKind::For(stmt) => stmt.eval(env, stdout),
            StmtKind::ReturnStmt(stmt) => stmt.eval(env, stdout),
            _ => unreachable!(),
        }
    }
}

impl Evaluable for ConditionStmt {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        let mut env = env.extend();
        let guard = self.condition.eval(&mut env, stdout)?;
        match guard {
            Value::Boolean(b) => {
                if b {
                    self.then_branch.eval(&mut env, stdout)
                } else {
                    match &self.else_branch {
                        None => Ok(Value::Void),
                        Some(stmt) => stmt.eval(&mut env, stdout),
                    }
                }
            }
            _ => Err(RuntimeError::ExpectedBoolean(
                self.condition.span(),
                guard.kind(),
            )),
        }
    }
}

impl Evaluable for WhileStmt {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        let cond = &self.condition;
        loop {
            let guard = cond.eval(env, stdout)?;
            match guard {
                Value::Boolean(b) => {
                    if b {
                        self.body.eval(env, stdout)?;
                    } else {
                        return Ok(Value::Void);
                    }
                }
                _ => return Err(RuntimeError::ExpectedBoolean(cond.span(), guard.kind())),
            }
        }
    }
}

impl Evaluable for ForStmt {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        let mut env = env.extend();
        let _ = &self.initializer.eval(&mut env, stdout)?;
        match &self.condition {
            None => {
                // infinite loop
                loop {
                    self.body.eval(&mut env, stdout)?;
                    self.increment.eval(&mut env, stdout)?;
                }
            }
            Some(cond) => loop {
                let guard = cond.eval(&mut env, stdout)?;
                match guard {
                    Value::Boolean(b) => {
                        if b {
                            self.body.eval(&mut env, stdout)?;
                            self.increment.eval(&mut env, stdout)?;
                        } else {
                            return Ok(Value::Void);
                        }
                    }
                    _ => return Err(RuntimeError::ExpectedBoolean(cond.span(), guard.kind())),
                }
            },
        }
    }
}

impl Evaluable for ReturnStmt {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        self.value
            .as_ref()
            .map_or(Err(RuntimeError::ReturnValue(Value::Void)), |expr| {
                let value = expr.eval(env, stdout)?;
                Err(RuntimeError::ReturnValue(value))
            })
    }
}

impl Evaluable for Option<Box<Stmt>> {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        if let Some(stmt) = self {
            stmt.eval(env, stdout)
        } else {
            Ok(Value::Void)
        }
    }
}

impl Evaluable for Option<Box<Expr>> {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        let val = match self {
            None => Value::Nil,
            Some(e) => e.eval(env, stdout)?,
        };
        Ok(val)
    }
}

impl Evaluable for Option<Expr> {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        let val = match self {
            None => Value::Nil,
            Some(e) => e.eval(env, stdout)?,
        };
        Ok(val)
    }
}

impl Evaluable for Expr {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        match &self.kind {
            ExprKind::Literal(l) => l.eval(env, stdout),
            ExprKind::Unary(u) => u.eval(env, stdout),
            ExprKind::Binary(b) => b.eval(env, stdout),
            ExprKind::Grouped(g) => g.eval(env, stdout),
            ExprKind::Variable(v) => v.eval(env, stdout),
            ExprKind::Assign(a) => a.eval(env, stdout),
            ExprKind::Call(c) => c.eval(env, stdout),
            _ => unreachable!(),
        }
    }
}

impl Evaluable for CallExpr {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        let callee = self.callee.eval(env, stdout)?;
        let args = self
            .args
            .iter()
            .map(|expr| expr.eval(env, stdout))
            .collect::<Result<Vec<Value>, RuntimeError>>()?;

        if let Value::Callable(callable) = callee {
            return match callable.call(args, env, stdout) {
                Ok(v) => Ok(v),
                Err(PrivateRuntimeError::BadArity { expected, found }) => {
                    Err(RuntimeError::BadArity {
                        callee_span: callable.params_span(),
                        caller_span: self.paren_token,
                        expected,
                        found,
                    })
                }
                Err(PrivateRuntimeError::Transparent(e)) => Err(e),
            };
        }

        Err(RuntimeError::CallOnNonCallable(
            self.callee.span(),
            callee.kind(),
        ))
    }
}

impl Evaluable for AssignExpr {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        let val = self.value.eval(env, stdout)?;
        env.mutate(&self.name, val.clone())?;
        Ok(val)
    }
}

impl Evaluable for BinaryExpr {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        let lhs = self.lhs.eval(env, stdout)?;
        let rhs = self.rhs.eval(env, stdout)?;
        let val = match &self.operator {
            op @ (BinaryOperator::EqualEqual | BinaryOperator::BangEq) => {
                return if lhs.kind() == rhs.kind()
                    && (lhs.kind() == ValueKind::Number
                        || lhs.kind() == ValueKind::Boolean
                        || lhs.kind() == ValueKind::String)
                {
                    match op {
                        BinaryOperator::EqualEqual => Ok(Value::Boolean(lhs == rhs)),
                        BinaryOperator::BangEq => Ok(Value::Boolean(lhs != rhs)),
                        _ => unreachable!(),
                    }
                } else {
                    Err(RuntimeError::IncompatibleBinaryOperand {
                        lhs_ty: lhs.kind(),
                        lhs_span: self.lhs.span(),
                        rhs_ty: rhs.kind(),
                        rhs_span: self.rhs.span(),
                    })
                }
            }
            BinaryOperator::GreaterEqual => {
                let lhs = inner_or!(
                    lhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.lhs.span(), "number")
                )?;
                let rhs = inner_or!(
                    rhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.rhs.span(), "number")
                )?;
                Value::Boolean(lhs >= rhs)
            }
            BinaryOperator::Greater => {
                let lhs = inner_or!(
                    lhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.lhs.span(), "number")
                )?;
                let rhs = inner_or!(
                    rhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.rhs.span(), "number")
                )?;
                Value::Boolean(lhs > rhs)
            }
            BinaryOperator::Less => {
                let lhs = inner_or!(
                    lhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.lhs.span(), "number")
                )?;
                let rhs = inner_or!(
                    rhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.rhs.span(), "number")
                )?;
                Value::Boolean(lhs < rhs)
            }
            BinaryOperator::LessEqual => {
                let lhs = inner_or!(
                    lhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.lhs.span(), "number")
                )?;
                let rhs = inner_or!(
                    rhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.rhs.span(), "number")
                )?;
                Value::Boolean(lhs <= rhs)
            }
            BinaryOperator::Minus => {
                let lhs = inner_or!(
                    lhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.lhs.span(), "number")
                )?;
                let rhs = inner_or!(
                    rhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.rhs.span(), "number")
                )?;
                Value::Number(lhs - rhs)
            }
            BinaryOperator::Plus => {
                let lhs = inner_or!(
                    lhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.lhs.span(), "number")
                )?;
                let rhs = inner_or!(
                    rhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.rhs.span(), "number")
                )?;
                Value::Number(lhs + rhs)
            }
            BinaryOperator::Slash => {
                let lhs = inner_or!(
                    lhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.lhs.span(), "number")
                )?;
                let rhs = inner_or!(
                    rhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.rhs.span(), "number")
                )?;
                Value::Number(lhs / rhs)
            }
            BinaryOperator::Star => {
                let lhs = inner_or!(
                    lhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.lhs.span(), "number")
                )?;
                let rhs = inner_or!(
                    rhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.rhs.span(), "number")
                )?;
                Value::Number(lhs * rhs)
            }
        };
        Ok(val)
    }
}

impl Evaluable for UnaryExpr {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        let value = self.expr.eval(env, stdout)?;
        match self.operator {
            UnaryOperator::Minus => {
                let num = inner_or!(
                    value,
                    number,
                    RuntimeError::InvalidUnaryOperand(self.expr.span(), "number")
                )?;
                Ok(Value::Number(-num))
            }
            UnaryOperator::Bang => {
                let bool = inner_or!(
                    value,
                    boolean,
                    RuntimeError::InvalidUnaryOperand(self.expr.span(), "boolean")
                )?;
                Ok(Value::Boolean(!bool))
            }
        }
    }
}

impl Evaluable for GroupedExpr {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        self.expr.eval(env, stdout)
    }
}

impl Evaluable for Identifier {
    fn eval(&self, env: &mut Environment, _stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        env.get(self)
    }
}

impl Evaluable for Literal {
    fn eval(
        &self,
        _env: &mut Environment,
        _stdout: &mut impl Write,
    ) -> Result<Value, RuntimeError> {
        let value = match *self {
            Literal::String(ref v) => Value::String(v.to_string()),
            Literal::Number(v) => Value::Number(v),
            Literal::Boolean(v) => Value::Boolean(v),
            Literal::Nil => Value::Nil,
        };
        Ok(value)
    }
}
