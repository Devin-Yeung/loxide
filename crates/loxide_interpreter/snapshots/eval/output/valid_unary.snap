---
source: crates/loxide_interpreter/src/eval.rs
expression: annotated_eval_test!(src)
---
print(-1.0  );
// => num(-1)
print(!true );
// => bool(false)
print(!false);
// => bool(true)
